### ISSUE TITLE
Method Not Allowed - NET6 Blazor WASM Hosted

### ISSUE THREAD

**Issue by jacarutech** [DuendeSoftware/Support#221]
**Which version of Duende BFF are you using?**
1.2.3
**Which version of .NET are you using?**
NET6
**Describe the bug**

This isn't really a bug, per se, and may not even be a Duende caused thing. I have just been struggling with this. I have a NET6 .NET hosted Blazor WASM project using Identity Server to manage user access and token mgmt. While developing everything works as expected. When releasing to an IIS server and trying to update content on the page I get a 405 response. I have looked at every solution I could before going the route of posting here.

System workflow:
- Page code call client service
- Client service makes api call to server backend
- Server controller generates token & calls server service
- Server service makes remote API call

**To Reproduce**

Make http request other than GET

**Expected behavior**

Process PUT, POST, DELETE request and route to backend server for handling.

**Log output/exception with stacktrace**

Images attached. File with "UserSet" are the headers I have set within the Program.cs file for responses. "PutCalled" are responses when I try a method other than GET.

**Additional context**

CLIENT Program.cs
`var builder = WebAssemblyHostBuilder.CreateDefault(args);
builder.RootComponents.Add<App>("#app");
builder.RootComponents.Add<HeadOutlet>("head::after");

builder.Services.AddAuthorizationCore();
builder.Services.AddScoped<AuthenticationStateProvider, BffAuthenticationStateProvider>();

builder.Services.AddTransient<AntiForgeryHandler>();
builder.Services.AddHttpClient("backend", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
    .AddHttpMessageHandler<AntiForgeryHandler>();
builder.Services.AddTransient(sp => sp.GetRequiredService<IHttpClientFactory>().CreateClient("backend"));

// Add additional specific services for the solution
builder.Services.AddAutoMapper(typeof(AutomapperProfiles).Assembly);
builder.Services.AddScoped<IAffiliateService, AffiliateService>();
builder.Services.AddSingleton<IAboutService, AboutService>();
builder.Services.AddSingleton<IServicesService, ServicesService>();
builder.Services.AddSingleton<ITechnologyService, TechnologyService>();
builder.Services.AddSingleton<IImageUploadService, ImageUploadService>();
builder.Services.AddMudServices(config =>
{
    config.SnackbarConfiguration.PositionClass = Defaults.Classes.Position.BottomRight;
    config.SnackbarConfiguration.VisibleStateDuration = 3000;
});

await builder.Build().RunAsync();
`
SERVER Program.cs
`var builder = WebApplication.CreateBuilder(args);

// Add identity services
var idsSettings = new IdentityServerSettings { ClientPassword = builder.Configuration["ClientSecrets:JacaruManagement"] };
#if DEBUG
idsSettings.DiscoveryUrl = "https://localhost:7102";
idsSettings.ClientName = "JacaruManagementDev";
#else
idsSettings.DiscoveryUrl = "https://identity.jacarumgmt.net";
idsSettings.ClientName = "JacaruManagement";
#endif

builder.Services.AddControllers();
builder.Services.AddRazorPages();
builder.Services.AddBff();

builder.Services.AddAuthentication(options =>
    {
        options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
        options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
        options.DefaultSignOutScheme = OpenIdConnectDefaults.AuthenticationScheme;
    })
    .AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, options =>
    {
        options.Cookie.SameSite = SameSiteMode.Lax;
    })
    .AddOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, options =>
    {
        options.Authority = idsSettings.DiscoveryUrl;
        options.ClientId = idsSettings.ClientName;
        options.ClientSecret = idsSettings.ClientPassword;
        options.ResponseType = OpenIdConnectResponseType.Code;
        options.ResponseMode = OpenIdConnectResponseMode.Query;
        options.MapInboundClaims = false;
        options.SaveTokens = true;
        options.GetClaimsFromUserInfoEndpoint = true;
        options.UsePkce = true;
        options.Scope.Clear();
        options.Scope.Add("JMS");
        options.Scope.Add("openid");
        options.Scope.Add("profile");
        options.Scope.Add("email");
        options.Scope.Add("offline_access");
    });

// Add services to the container.

builder.Services.AddAutoMapper(typeof(AutomapperProfiles).Assembly);
builder.Services.AddSettingsServiceConfigurations(builder.Configuration);
builder.Services.AddServicesInjectors();

#if DEBUG
ApiEndpoints.ApiBaseUrl = new Uri("https://localhost:7200");
#else
ApiEndpoints.ApiBaseUrl = new Uri("https://gateway.jacarumgmt.net");
#endif

builder.Services.AddHttpClient("JmsClient", options =>
{
    options.BaseAddress = ApiEndpoints.ApiBaseUrl;
    options.DefaultRequestHeaders.CacheControl = new CacheControlHeaderValue { MaxAge = TimeSpan.FromSeconds(2592000) };
}).AddUserAccessTokenHandler();

builder.Services.AddHttpContextAccessor();

builder.Logging.SetMinimumLevel(LogLevel.Error);

var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseWebAssemblyDebugging();
}
else
{
    app.UseExceptionHandler("/Error");
    app.Use(async (context, next) =>
    {
        context.Response.Headers.XXSSProtection = "1; mode=block";
        context.Response.Headers.XFrameOptions = "SAMEORIGIN";
        context.Response.Headers.XContentTypeOptions = "nosniff";
        context.Response.Headers.Add("Referrer-Policy", "strict-origin-when-cross-origin");
        context.Response.Headers.Allow = "GET, POST, DELETE, PUT, OPTIONS";

        context.Response.Headers.ContentSecurityPolicy =
            "default-src 'self'; " +
            "frame-ancestors 'none'; " +
            "font-src 'self' https://fonts.googleapis.com https://fonts.gstatic.com https://cdn.jsdelivr.net https://cdnjs.cloudflare.com; " +
            "style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net https://cdnjs.cloudflare.com https://fonts.googleapis.com https://fonts.gstatic.com; " +
            "script-src 'self' 'unsafe-eval' 'unsafe-inline' https://cdn.jsdelivr.net https://use.fontawesome.com https://www.google.com https://maps.googleapis.com https://www.gstatic.com; " +
            "img-src 'self' data: https://www.google.com https://maps.googleapis.com https://www.gstatic.com https://maps.gstatic.com; " +
            "connect-src 'self' https://cdn.jsdelivr.net https://cdnjs.cloudflare.com https://use.fontawesome.com https://maps.googleapis.com https://www.google.com https://fonts.googleapis.com https://fonts.gstatic.com https://www.gstatic.com; " +
            "frame-src https://www.google.com https://maps.googleapis.com https://www.gstatic.com;";

        await next();
    });
}

app.UseHttpsRedirection();

app.UseBlazorFrameworkFiles();
app.UseStaticFiles();

app.UseRouting();
app.UseAuthentication();
app.UseBff();
app.UseAuthorization();

app.MapBffManagementEndpoints();
app.MapRazorPages();
app.MapControllers()
    .RequireAuthorization()
    .AsBffApiEndpoint();
app.MapFallbackToFile("index.html");

app.Run();`

CONTROLLER Example
`[ApiController]
    [Route("api/[controller]")]
    public class BaseApiController : ControllerBase
    {
    }`
`public class JacaruTechController : BaseApiController
    {
        private readonly IAboutService _aboutService;
        private readonly IServicesService _servicesService;
        private readonly ITechnologyService _technologyService;
        private readonly ILogger<JacaruTechController> _logger;

        public JacaruTechController(IAboutService aboutService, IServicesService servicesService, ITechnologyService technologyService, ILogger<JacaruTechController> logger)
        {
            _aboutService = aboutService;
            _servicesService = servicesService;
            _technologyService = technologyService;
            _logger = logger;
        }

        #region About

        [HttpGet("about")]
        public async Task<ServiceResponse<List<SectionDto>>> GetAboutSections()
        {
            try
            {
                return await _aboutService.GetAboutSectionsAsync();
            }
            catch (Exception e)
            {
                _logger.LogError(e.Message);
                throw;
            }
        }

        [HttpDelete("about/{id:int}")]
        public async Task<ServiceResponse<SectionDto>> DeleteAboutSection(int id)
        {
            var token = await GenerateTokenAsync();

            var response = await _aboutService.DeleteAboutSectionAsync(id, token);

            return response;
        }
}`
![ResponseHeadersUserSet](https://user-images.githubusercontent.com/85734366/191000265-9456664e-ae6
![ResponseHeadersPutCalled](https://user-images.githubusercontent.com/85734366/191000277-81f070ad-01d6-47d5-957f-92ffad720b32.png)
3-4107-823c-5e6c87c8eb0e.png)

**Comment by jacarutech** [comment]
![ResponseHeadersUserSet](https://user-images.githubusercontent.com/85734366/191000582-e96eed3d-d72e-4ac0-b304-5f5ea921f0bd.png)

**Comment by brockallen** [comment]
Sorry, despite all of the code pasted above, I'm having a hard time trying to understand what request where you're having a problem with. I think you'll have to track down exactly which request (e.g. is it from the front-end to the back-end, or is it from the back-end to the remote API, or is it something somewhere else) for us to help.

**Comment by jacarutech** [comment]
The issue is in the call from the Frontend to the backend PUT api. I get a 405 - Method Not Allowed error. Response headers show allowed methods are GET, HEAD, OPTIONS, TRACE. I've checked my gateway logs and it's not making it through to the gateway, which is the next stop, so it's got to be happening between the WASM client frontend and the hosted backend (which uses IDS BFF framework).

**Comment by jacarutech** [comment]
For additional context, the call is routed to an external Ocelot API Gateway with the payload and token.

**Comment by brockallen** [comment]
> The issue is in the call from the Frontend to the backend PUT api.

Then this should be fine -- sounds like maybe like there's no routing configured for that path that allows PUT? But all in all, sounds unrelated to BFF -- more just an ASP.NET issue?

**Comment by jacarutech** [comment]
It may be ASP and not BFF, that's why I am throwing things at the wall to see if anything sticks. The route is correct as it works in dev environment. I do repeatedly get this error in my IDS logs. Could this have something to do with it?

**Comment by jacarutech** [comment]
[14:04:22 Error] Duende.IdentityServer.Services.KeyManagement.KeyManager
Error unprotecting key with kid <<HIDDEN BY USER>>.
System.Security.Cryptography.CryptographicException: The key {<<HIDDEN BY USER>>} was not found in the key ring. For more information go to http://aka.ms/dataprotectionwarning

**Comment by brockallen** [comment]
> It may be ASP and not BFF, that's why I am throwing things at the wall to see if anything sticks. The route is correct as it works in dev environment. I do repeatedly get this error in my IDS logs. Could this have something to do with it?

Hmm, then maybe it's something about your prod environment -- perhaps some firewall or reverse proxy setting, perhaps?

**Comment by brockallen** [comment]
> [14:04:22 Error] Duende.IdentityServer.Services.KeyManagement.KeyManager
Error unprotecting key with kid <>.
System.Security.Cryptography.CryptographicException: The key {<>} was not found in the key ring. For more information go to http://aka.ms/dataprotectionwarning

This is unrelated -- likely due to misconfigured data protection on the prod instance: https://docs.duendesoftware.com/identityserver/v6/deployment/data_protection/

**Comment by jacarutech** [comment]
I have reached out to my web server host to get their input. Thank you for the lead on data protection. Thanks for taking some time to try to help me figure this out.

**Comment by brockallen** [comment]
Any updates?

**Comment by jacarutech** [comment]
After struggling for a while, and having a convo with the server team, the solution was a server issue. Even though I tried this, I had to tweak this answer to get it working: https://stackoverflow.com/questions/9854602/asp-net-web-api-405-http-verb-used-to-access-this-page-is-not-allowed-

I couldn't add the remove handler part to the config without the server barking at me. Just adding the remove module part worked. Now if I could just get a production signing credential working properly I would be all set :-) The joys of teaching yourself coding and cryptography.
